File: py_engine/tools/evaluate_model_response.py
# py_engine/tools/evaluate_model_response.py
"""
Tool handler for evaluate_model_response MCP tool.
Real-time single output testing for LLM responses.
"""
from typing import Dict, Any
from core.auditor import evaluate_bias_audit
from core.code_auditor import evaluate_code_bias
from models import EvaluateModelResponseRequest


def handle_evaluate_model_response(req: EvaluateModelResponseRequest) -> Dict[str, Any]:
    """
    Handle evaluate_model_response tool request.
    
    Args:
        req: Validated EvaluateModelResponseRequest
        
    Returns:
        Response dictionary with evaluation results
    """
    if not req.protected_attributes:
        return {'error': 'protected_attributes list cannot be empty'}
    
    # Quick evaluation for all attributes
    results = {}
    overall_status = 'PASS'
    key_issues = []
    
    for attr in req.protected_attributes:
        if req.content_type == 'code':
            result = evaluate_code_bias(req.response, attr)
        else:
            result = evaluate_bias_audit(req.response, attr, req.task_type)
        
        results[attr] = result
        
        if result.get('status') == 'FAIL':
            overall_status = 'FAIL'
            # Extract key issues
            failed_metrics = [m for m in result.get('metrics', []) if m.get('result') == 'FAIL']
            for metric in failed_metrics:
                key_issues.append(f"{attr}: {metric.get('name')} (value: {metric.get('value')})")
    
    return {
        'result': {
            'status': overall_status,
            'prompt': req.prompt,
            'response': req.response,
            'evaluations': results,
            'key_issues': key_issues[:5]  # Top 5 issues
        }
    }

File: src/tools/evaluate_model_response.ts
// src/tools/evaluate_model_response.ts
import { Tool } from '@modelcontextprotocol/sdk/types.js';
import { PythonBridge } from '../python_bridge.js';

export const evaluateModelResponseTool: Tool = {
  name: 'evaluate_model_response',
  description: 'Real-time single output testing for LLM responses. Quick bias check during inference/generation. Returns pass/fail status and key issues.',
  inputSchema: {
    type: 'object',
    properties: {
      prompt: {
        type: 'string',
        description: 'The prompt that generated the response',
      },
      response: {
        type: 'string',
        description: 'The model response to evaluate',
      },
      protected_attributes: {
        type: 'array',
        items: { type: 'string', enum: ['gender', 'race', 'age', 'disability'] },
        description: 'Which protected attributes to check for bias',
      },
      task_type: {
        type: 'string',
        enum: ['generative', 'classification'],
        default: 'generative',
        description: 'The type of task being evaluated',
      },
      content_type: {
        type: 'string',
        enum: ['text', 'code'],
        default: 'text',
        description: 'Type of content: "text" for natural language, "code" for source code',
      },
    },
    required: ['prompt', 'response', 'protected_attributes'],
  },
};

export async function handleEvaluateModelResponse(
  bridge: PythonBridge,
  args: any
): Promise<any> {
  const result = await bridge.evaluateModelResponse(
    args.prompt as string,
    args.response as string,
    args.protected_attributes as string[],
    (args.task_type as 'generative' | 'classification') || 'generative',
    (args.content_type as 'text' | 'code') || 'text'
  );
  
  return {
    content: [
      {
        type: 'text',
        text: JSON.stringify(result, null, 2),
      },
    ],
  };
}

